oonib/testhelpers/http_helpers.py

```diff
@@ -168,7 +241,336 @@ def all(self, length, keyword):
+
+    @defer.inlineCallbacks
+    def tcp_connect(self, socket):
+        cached_value = yield self.lookup('tcp_connect', socket)
+        if cached_value is not None:
+            defer.returnValue(cached_value)
+
+        socket_info = {
+            'status': None,
+            'failure': None
+        }
+
+        ip_address, port = socket.split(":")
+        try:
+            point = TCP4ClientEndpoint(reactor, ip_address, int(port))
+            yield point.connect(TCPConnectFactory())
+            socket_info['status'] = True
+        except TimeoutError:
+            socket_info['status'] = False
+            socket_info['failure'] = 'generic_timeout_error'
+        except ConnectionRefusedError:
+            socket_info['status'] = False
+            socket_info['failure'] = 'connection_refused_error'
+        except:
+            socket_info['status'] = False
+            socket_info['failure'] = 'unknown_error'
+        yield self.cache_value('tcp_connect', socket, socket_info)
+        defer.returnValue(socket_info)
+
+    @defer.inlineCallbacks
+    def dns_consistency(self, hostname):
+        cached_value = yield self.lookup('dns_consistency', hostname)
+        if cached_value is not None:
+            defer.returnValue(cached_value)
+
+        dns_info = {
+            'addrs': [],
+            'failure': None
+        }
+
+        try:
+            records = yield dns_client.lookupAddress(hostname)
+            answers = records[0]
+            for answer in answers:
+                if answer.type is dns.A:
+                    dns_info['addrs'].append(answer.payload.dottedQuad())
+                elif answer.type is dns.CNAME:
+                    dns_info['addrs'].append(answer.payload.name.name)
+        except DNSNameError:
+            dns_info['failure'] = 'dns_name_error'
+        except DNSServerError:
+            dns_info['failure'] = 'dns_server_failure'
+        except:
+            dns_info['failure'] = 'unknown_error'
+
+        yield self.cache_value('dns_consistency', hostname, dns_info)
+        defer.returnValue(dns_info)
+
+
+web_connectivity_cache = WebConnectivityCache()
+
+class WebConnectivity(RequestHandler):
+    @defer.inlineCallbacks
+    def control_measurement(self, http_url, socket_list):
+        hostname = urlparse(http_url).netloc
+        dl = [
+            web_connectivity_cache.http_request(http_url),
+            web_connectivity_cache.dns_consistency(hostname)
+        ]
+        for socket in socket_list:
+            dl.append(web_connectivity_cache.tcp_connect(socket))
+        responses = yield defer.DeferredList(dl)
+        http_request = responses[0][1]
+        dns = responses[1][1]
+        tcp_connect = {}
+        for idx, response in enumerate(responses[2:]):
+            tcp_connect[socket_list[idx]] = response[1]
+        self.finish({
+            'http_request': http_request,
+            'tcp_connect': tcp_connect,
+            'dns': dns
+        })
```

Comment:
Running the backend in a local Vagrant and using `curl(1)` as follows:

```
curl -vX POST --data @INPUT http://127.0.0.1:57007/
```

where `INPUT` contains

```
{"http_request": "https://ooni.torproject.org/","tcp_connect":["154.35.132.70","82.195.75.101","38.229.72.16","138.201.14.197","86.59.30.40"]}
```

it yields:

```
* About to connect() to 127.0.0.1 port 57007 (#0)
*   Trying 127.0.0.1... connected
> POST / HTTP/1.1
> User-Agent: curl/7.22.0 (i686-pc-linux-gnu) libcurl/7.22.0 OpenSSL/1.0.1 zlib/1.2.3.4 libidn/1.23 librtmp/2.3
> Host: 127.0.0.1:57007
> Accept: */*
> Content-Length: 142
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 142out of 142 bytes
```

and then `curl(1)` is stuck more or less forever (depending on how you define "forever").

Now, looking at the `oonib` side, one sees this in the logs:

```
--- <exception caught here> ---
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 1128, in _inlineCallbacks
    result = g.send(result)
  File "/data/oonib/oonib/testhelpers/http_helpers.py", line 478, in tcp_connect
    ip_address, port = socket.split(":")
exceptions.ValueError: need more than 1 value to unpack
```

This error is caused by the fact that I supplied wrong input (the port was not specified). But what is interesting is that no error is ever returned to the client.

I believe this happens because `@synchronous` tells cyclone that `finish` must explicitly be called whilean exception is raised when calling `web_connectivity_cache.tcp_connect` and hence the final `finish` function is never called.

If I'm right, the proper fix would probably be to wrap the body of `control_measurement` with `try..catch` and to make sure that `self.finish()` is called also on exception -- or, possibly, as part of `finally`.